mm/pmm.c

```diff
@@ -421,6 +421,83 @@ static frame_t *find_mfn_frame(list_head_t *list, mfn_t mfn, unsigned int order)
     return NULL;
 }
 
+frame_t *find_free_mfn_frame(mfn_t mfn, unsigned int order) {
+    frame_t *frame;
+
+    spin_lock(&lock);
+    frame = _find_mfn_frame(free_frames, mfn, order);
+    spin_unlock(&lock);
+
+    return frame;
+}
+
+frame_t *find_busy_mfn_frame(mfn_t mfn, unsigned int order) {
+    frame_t *frame;
+
+    spin_lock(&lock);
+    frame = _find_mfn_frame(busy_frames, mfn, order);
+    spin_unlock(&lock);
+
+    return frame;
+}
+
+frame_t *find_mfn_frame(mfn_t mfn, unsigned int order) {
+    frame_t *frame;
+
+    spin_lock(&lock);
+    frame = _find_mfn_frame(busy_frames, mfn, order);
+    if (!frame)
+        frame = _find_mfn_frame(free_frames, mfn, order);
+    spin_unlock(&lock);
```

Comment:
Could we re-use `find_{busy,free}_mfn_frame` here? Or is it a bad idea with the lock-unlock-lock-unlock behavior imposed?

Reply:
Yeah, it was done like that to avoid locking-unlocking problems. We now will have a static function to be used internally by PMM without locking and a bunch of public interfaces which handle the locking.